src/draw/sw/lv_draw_sw_rect.c

```diff
@@ -154,23 +328,54 @@ LV_ATTRIBUTE_FAST_MEM static void draw_bg(const lv_area_t * coords, const lv_are
 
     /*In case of horizontal gradient pre-compute a line with a gradient*/
     lv_color_t * grad_map = NULL;
-    lv_color_t * grad_map_ofs = NULL;
-    if(grad_dir == LV_GRAD_DIR_HOR) {
+    lv_coord_t off_x = 0;
+#if !DITHER_GRADIENT
+    if(grad_dir != LV_GRAD_DIR_NONE) {
         grad_map = lv_mem_buf_get(coords_w * sizeof(lv_color_t));
         int32_t i;
         for(i = 0; i < coords_w; i++) {
             grad_map[i] = grad_get(dsc, coords_w, i);
         }
-        grad_map_ofs = grad_map;
-        if(dsc->bg_grad_dir == LV_GRAD_DIR_HOR) grad_map_ofs += draw_area.x1 - coords_bg.x1;
+        if(dsc->bg_grad_dir == LV_GRAD_DIR_HOR) off_x = draw_area.x1 - coords_bg.x1;
     }
-
+#else
+    lv_coord_t grad_size = grad_dir == LV_GRAD_DIR_HOR ? coords_w : coords_h;
+    lv_color32_t * hires_grad_map = NULL;
+    scolor24_t * error_acc = NULL;
+    if(grad_dir != LV_GRAD_DIR_NONE) {
+        /*Precompute the gradient perfect color in 24bits*/
+        hires_grad_map = lv_mem_buf_get(grad_size * sizeof(lv_color32_t));
+        for(int32_t i = 0; i < grad_size; i++) {
+            hires_grad_map[i] = hires_grad_get(dsc, grad_size, i);
+        }
+        /*Used to store current line gradient after dithering*/
+        grad_map = lv_mem_buf_get(coords_w * sizeof(lv_color_t));
+        if(dsc->bg_grad_dir == LV_GRAD_DIR_HOR) off_x = draw_area.x1 - coords_bg.x1;
+        /*Used to store error to apply on the next line for dithering*/
+        error_acc = lv_mem_buf_get(coords_w * sizeof(scolor24_t));
+        lv_memset_00(error_acc, coords_w * sizeof(scolor24_t));
+    }
+#endif
     int32_t h;
     lv_draw_mask_res_t mask_res;
     lv_area_t blend_area;
+    blend_func_t blend_func = NULL;
     blend_area.x1 = draw_area.x1;
     blend_area.x2 = draw_area.x2;
 
+
+    if(grad_dir == LV_GRAD_DIR_NONE)
+        blend_func = &blend_fill_line;
+    else if(grad_dir == LV_GRAD_DIR_HOR)
+        blend_func = &blend_map_line;
+    else if(grad_dir == LV_GRAD_DIR_VER) {
+#if DITHER_GRADIENT
+        blend_func = &blend_map_line;
+#else
+        blend_func = &blend_fill_vert_line;
+#endif
```

Comment:
I'm working on improving the architecture of the drawing part and made a similar simplification. https://github.com/lvgl/lvgl/blob/699cf2d63cb477f9ec76a788a4be26b9e60c4577/src/draw/sw/lv_draw_sw_rect.c#L184-L190

Reply:
Not sure I'm following you here on your refactor. `lv_draw_sw_blend` as I'm reading it is not a pointer to the right function. It's a pseudo generic function that'll contain something like:
```
if (some mode) do something
else if (some other mode) do some other thing 
else ...
```
Since it's inside a for loop, and it's a function that's not visible to the compiler (different compile unit), it can't be unrolled, so it'll be very heavy (for each iteration, you'll have to pay for the same tests again and again). IIUC, you're hiding the mess of the *current* code under the rug (in its own file). 

It's good for the readability of the code through.

In my solution above, I'm setting the "function" to use depending on some flags that are uniform/constant for the whole loop. It allows having a different line drawing function based on if the gradient is horizontal, vertical, with dithering, etc...

It's a shame that it's C code. In C++, you'd write a lambda and have the compiler write the optimal code for you here without the cost of calling a function. Here, in both your and my solution, we have to pay for calling a function per line.
In your solution, you also pay for finding out what to do in that function (with tests) while in mine, I'm not paying this cost.

Reply:
Yes, it's only a wrapper to use a unified function. Basically, we moved the `if` from `lv_draw_rect` to `lv_draw_sw_blend`.

I think this extra if doesn't matter much compared to the overall operation to render the UI.